### PR TITLE
Feature/Xline client kv

### PR DESCRIPTION
**Pull request by bsbds**
Implement kv client in `xline-client`

based on #317

Please briefly answer these questions:

* what problem are you trying to solve? (or if there's no problem, what's the motivation for this change?)

* what changes does this pull request make?

* are there any non-obvious implications of these changes? (does it break compatibility with previous versions, etc)

### CODE REVIEW

File: xline-client/src/clients/kv.rs [hunk 1]
```diff
@@ -27,4 +31,762 @@ impl KvClient {
...
+    }
+
+    /// Takes a list of comparison. If all comparisons passed in succeed,
+    /// the operations passed into `and_then()` will be executed. Or the operations
+    /// passed into `or_else()` will be executed.
+    ///
+    /// # Panics
+    ///
+    /// panics if `when` is called twice or called after `when` or called after `or_else`
+    #[inline]
```
Comment (iGxnon) [inline xline-client/src/clients/kv.rs:728]: I guess we could remove `#[inline]` here according to [this](https://matklad.github.io/2021/07/09/inline-in-rust.html#Inlining-in-Practice)
> Fifth, mind generic functions. It’s not too wrong to say that generic functions are implicitly inline. As a result, they often are a cause for code bloat. Generic functions, especially in libraries, should be written to minimize unwanted inlining. To give an example from [wat](https://github.com/bytecodealliance/wasm-tools/blob/0486fb4de505b8116a0034bdde4918cd783325b9/crates/wat/src/lib.rs#L214-L222):

File: xline-client/src/clients/kv.rs [hunk 2]
```diff
@@ -27,4 +31,762 @@ impl KvClient {
...
+        self
+    }
+
+    /// Takes a list of operations. The operations list will be executed, if the
+    /// comparisons passed in `when()` succeed.
+    ///
+    /// # Panics
+    ///
+    /// panics if `and_then` is called twice or called after `or_else`
+    #[inline]
```
Comment (iGxnon) [inline xline-client/src/clients/kv.rs:747]: same as above

File: xline-client/src/clients/kv.rs [hunk 3]
```diff
@@ -27,4 +31,762 @@ impl KvClient {
...
+        self
+    }
+
+    /// Takes a list of operations. The operations list will be executed, if the
+    /// comparisons passed in `when()` fail.
+    ///
+    /// # Panics
+    ///
+    /// panics if `or_else` is called twice
+    #[inline]
```
Comment (iGxnon) [inline xline-client/src/clients/kv.rs:770]: same as above

File: xline-client/src/clients/kv.rs (hunk 1 above)
Comment (bsbds) [reply xline-client/src/clients/kv.rs:728]: The clippy requires inline in public functions. As mentioned in the article, the generic functions are implicitly inlined, adding `#[inline]` just explicitly does it. I think what the author means is to add method delegation in generic functions, but that sounds like premature optimization?

Comment (iGxnon) [reply xline-client/src/clients/kv.rs:728]: OK, if clippy doesn't allow it, then just keep it.

File: xline-client/tests/kv.rs [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+
+        assert!(!resp.succeeded);
+        let op_responses = resp.responses;
+        assert_eq!(op_responses.len(), 1);
+
+        match op_responses[0].response.as_ref().unwrap() {
+            xlineapi::Response::ResponseRange(resp) => {
+                assert_eq!(resp.kvs[0].value, b"02")
+            }
+            _ => panic!("unexpected response"),
```
Comment (rogercloud) [inline xline-client/tests/kv.rs:211]: please tell what's expected here, and all the "panic unexpected response". Yes it tells the truth, but it's helpless when debugging the test code.

File: xline-client/src/clients/kv.rs [hunk 5]
```diff
@@ -27,4 +31,762 @@ impl KvClient {
...
+    #[must_use]
+    pub fn when(mut self, compares: impl Into<Vec<Compare>>) -> Self {
+        assert!(!self.c_when, "cannot call when twice");
+        assert!(!self.c_then, "cannot call when after and_then");
+        assert!(!self.c_else, "cannot call when after or_else");
+
+        let compares_vec: Vec<Compare> = compares.into();
+        self.c_when = true;
+        self.req.compare = compares_vec.iter().map(|c| c.clone().0).collect();
```
Comment (themanforfree) [inline xline-client/src/clients/kv.rs:737]: Please use `into_iter` then we don't need to clone it.

File: xline-client/tests/kv.rs [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+        .await?
+        .kv_client();
+    let request = PutRequest::new("put", "123");
+    client.put(request).await?;
+
+    // overwrite with prev key
+    {
+        let request = PutRequest::new("put", "456").with_prev_kv(true);
+        let resp = client.put(request).await?;
+        let prev_key = resp.prev_kv;
```
Comment (themanforfree) [inline xline-client/tests/kv.rs:23]: `prev_kv`, not `prev_key`

File: xline-client/tests/kv.rs [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+        assert_eq!(resp.kvs[1].key, b"get11");
+        assert_eq!(resp.kvs[1].value, b"11");
+    }
+
+    Ok(())
+}
+
+#[tokio::test]
+async fn test_delete() -> Result<()> {
+    let mut cluster = Cluster::new(3).await;
```
[the remark this answers is not in the document]
Comment (bsbds) [reply xline-client/tests/kv.rs:99]: done

File: xline-client/tests/kv.rs (hunk 4 above)
Comment (bsbds) [reply xline-client/tests/kv.rs]: done

File: xlineapi/src/lib.rs [hunk 8]
```diff
@@ -606,6 +606,11 @@ impl RequestWithToken {
             request,
         }
     }
+
+    /// New `RequestWithToken` with token
+    pub fn new_with_option_token(request: RequestWrapper, token: Option<String>) -> Self {
+        RequestWithToken { token, request }
+    }
 }
```
[the remark this answers is not in the document]
Comment (bsbds) [reply xlineapi/src/lib.rs]: done

File: xline-client/tests/kv.rs (hunk 6 above)
Comment (bsbds) [reply xline-client/tests/kv.rs]: done

File: xline-client/src/clients/kv.rs (hunk 5 above)
Comment (bsbds) [reply xline-client/src/clients/kv.rs]: done

File: xline/src/server/auth_server.rs [hunk 9]
```diff
@@ -86,7 +86,9 @@ where
         T: Into<RequestWrapper>,
     {
         let wrapper = match get_token(request.metadata()) {
-            Some(token) => RequestWithToken::new_with_token(request.into_inner().into(), token),
+            Some(token) => {
+                RequestWithToken::new_with_token(request.into_inner().into(), Some(token))
+            }
             None => RequestWithToken::new(request.into_inner().into()),
         };
```
Comment (themanforfree) [inline xline/src/server/auth_server.rs:93]:
```rust
let token = get_token(request.metadata());
let wrapper = RequestWithToken::new_with_token(request.into_inner().into(), token)
```

File: xline/src/server/kv_server.rs [hunk 10]
```diff
@@ -135,7 +135,9 @@ where
         T: Into<RequestWrapper> + Debug,
     {
         let wrapper = match get_token(request.metadata()) {
-            Some(token) => RequestWithToken::new_with_token(request.into_inner().into(), token),
+            Some(token) => {
+                RequestWithToken::new_with_token(request.into_inner().into(), Some(token))
+            }
             None => RequestWithToken::new(request.into_inner().into()),
         };
```
Comment (themanforfree) [inline xline/src/server/kv_server.rs:142]: Same as above

File: xline/src/server/kv_server.rs [hunk 11]
```diff
@@ -403,7 +405,9 @@ where
         Self::check_range_request(range_req, self.kv_storage.revision())?;
         let is_serializable = range_req.serializable;
         let wrapper = match get_token(request.metadata()) {
-            Some(token) => RequestWithToken::new_with_token(request.into_inner().into(), token),
+            Some(token) => {
+                RequestWithToken::new_with_token(request.into_inner().into(), Some(token))
+            }
             None => RequestWithToken::new(request.into_inner().into()),
         };
```
Comment (themanforfree) [inline xline/src/server/kv_server.rs:412]: Same as above

File: xline/src/server/lease_server.rs [hunk 12]
```diff
@@ -135,7 +135,9 @@ where
         T: Into<RequestWrapper>,
     {
         let wrapper = match get_token(request.metadata()) {
-            Some(token) => RequestWithToken::new_with_token(request.into_inner().into(), token),
+            Some(token) => {
+                RequestWithToken::new_with_token(request.into_inner().into(), Some(token))
+            }
             None => RequestWithToken::new(request.into_inner().into()),
         };
```
Comment (themanforfree) [inline xline/src/server/lease_server.rs:142]: Same as above

File: xline/src/server/lock_server.rs [hunk 13]
```diff
@@ -90,7 +90,7 @@ impl LockServer {
         T: Into<RequestWrapper>,
     {
         let wrapper = match token {
-            Some(token) => RequestWithToken::new_with_token(request.into(), token),
+            Some(token) => RequestWithToken::new_with_token(request.into(), Some(token)),
             None => RequestWithToken::new(request.into()),
         };
```
Comment (themanforfree) [inline xline/src/server/lock_server.rs:95]: Same as above

File: xline/src/server/auth_server.rs (hunk 9 above)
Comment (bsbds) [reply xline/src/server/auth_server.rs]: done

**Review by themanforfree — APPROVED** [review]

**Review by Phoenix500526 — APPROVED** [review]

**Pull request merged**